### ISSUE TITLE
Relationship to core web vitals

### ISSUE THREAD

**Issue by dvoytenko** [domenic/cooperatively-sized-iframes#4]
It's worth discussing this feature in the context of web vitals. For instance, this proposal may make CLS more likely. 

IMHO this proposal is a good idea. It'd just be good to define a few recommendations of how to ensure good LCP and CLS. For instance, defining `min-height` could help achieve stable CLS. Etc.

**Comment by domenic** [comment]
+1 to considering this carefully, especially in web-developer-facing examples and guidance.

To expand, I think the main use case would be something like:

* I want to embed an iframe (e.g., a tweet)
* I know the width will always be a specific number
* I want to ensure we don't clip the content (i.e., have iframe height < requested height). If this causes layout shifts, so be it.
* But, I am OK with some extra vertical padding, if it avoids layout shifts.

Let's make this concrete: let's say

* Tweets always have width 500px
* Tweets generally range between height 200px-400px. But they can sometimes go higher or lower.
* We have two example tweets: tweet1 is 350px high, and tweet2 is 420px high.

The *bad* way to do this would be:

```html
<iframe src="..." style="contain-intrinsic-size: from-element"></iframe>
```

This would cause the dimensions to start out at 300x150, and then snap to the correct value (500x350 or 500x420). Lots of layout shift. Sad.

The *good* way to do this would be:

```html
<iframe src="..."
        style="contain-intrinsic-size: from-element 500px 400px;
               min-height: 400px;"></iframe>
```

This would cause the dimensions to start out at 500x400, and:

- In the case of tweet1, stay at those dimensions and cause no layout shift. (But end up with 50px of extra vertical whitespace.)
- In the case of tweet2, cause 20px of layout shift as the iframe expands from 400px to 420px.

**Comment by domenic** [comment]
We should figure out how to allow the inner frame to have similar requests, e.g. "between 400px and 600px height, but I prefer 432px". (If you are at 700px does this clamp to 600px to minimize layout shift, or does it say "eh, I have layout shift coming anyway, just go to the correct 432px"? Maybe we want to allow either.)

This will require a more complicated microsyntax than the current `requestedwidth=""`/`requestedheight=""`.

**Comment by domenic** [comment]
(In what follows I'll talk only about width; height is symmetrical.)

Goals for outer page:

- Express max-width/min-width which interact with requested width
- Express width override which causes requested width to be ignored
- Express default or initial width

CSS seems good for this. We're using `max-width`/`min-width`/`width`/`contain-intrinsic-width`'s fallback value, respectively.

Goals for inner page:

- Express (min requested width, preferred requested width, max requested width)
- Express strategy for if current width starts or becomes outside of the requested width range. Two possibilities:
   - If current width is outside the range, clamp to the range (i.e., minimize layout shift)
   - If current width is outside the range, go straight to preferred width (eat the layout shift all at once to get the optimal size)

One idea for a proposal:

```html
<meta name="requested-viewport" value="width=500, min-width=400, max-width=600, width-mode=clamp">
```

Modes are "clamp" and "jump", default "clamp".

Omitting `min-width` or `max-width` makes them default to `width`.

Not really sure how to spec the computation algorithm, given all these inputs. Basically, we send all the inputs into the parent process, so that there we have:
- `requested-width`, `requested-min-width`, `requested-max-width` (all numbers of pixels), `requested-width-mode` (clamp or jump)
- `contain-intrinsic-width`, `max-width`, `min-width`, `width` (CSS properties)
- current width

Seems like something the smart people who already figured out how `width`/`max-width`/`min-width`/`contain-intrinsic-width` interact should be able to help with though :). /cc @tabatkins

**Comment by dvoytenko** [comment]
What does `width-mode=jump` do vs `clamp`? Otherwise this proposal makes sense to me. However, would the manual change of this `<meta>` value also rerequest the new size from parent? Or would the iframe page rely on `window.resizeTo()` API?

**Comment by domenic** [comment]
- `width-mode=jump`: If current width is outside the range, go straight to preferred width (eat the layout shift all at once to get the optimal size)

- `width-mode=clamp`: If current width is outside the range, clamp to the range (i.e., minimize layout shift)

It's not clear we need both in v1; we could pick one (probably clamp?) as the default and not have a `width-mode` for now. But I wanted to make sure we could add it in the future.

> However, would the manual change of this `<meta>` value also rerequest the new size from parent?

Yes.

**Comment by dvoytenko** [comment]
Makes sense. And the "current" width is defined as `documentElement.offsetWidth`? Or `documentElement.scrollWidth`?

Overall the behavior described by `width-mode` makes sense. But I think it might be more of a parent page's responsibility. The iframe page can request whatever it wants. But the parent page has to deal with layout shift and other tradeoffs. As a parent, the page might decide to be more careful with iframes in the top of the documents (higher chance of layout shift disruption), and allow more to iframes at the bottom of the document (almost no chance of layout shift disruption).

**Comment by domenic** [comment]
> And the "current" width is defined as documentElement.offsetWidth? Or documentElement.scrollWidth?

I think scrollWidth is best? But I admit not having thought about it in detail; this is another area where the aforementioned smart CSS people can help point the right way.

> But I think it might be more of a parent page's responsibility.

That's a good point.
